File: WowUp.WPF/Services/WarcraftService.cs
using ProtoBuf;
using Serilog;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using WowUp.Common.Enums;
using WowUp.Common.Extensions;
using WowUp.Common.Models;
using WowUp.Common.Models.Events;
using WowUp.Common.Models.Warcraft;
using WowUp.WPF.Entities;
using WowUp.WPF.Models.WowUp;
using WowUp.WPF.Repositories.Contracts;
using WowUp.WPF.Services.Contracts;
using WowUp.WPF.Utilities;

namespace WowUp.WPF.Services
{
    public class WarcraftService : IWarcraftService
    {
        // WOW STRINGS
        private const string ClassicFolderName = "_classic_";
        private const string ClassicPtrFolderName = "_classic_ptr_";
        private const string RetailFolderName = "_retail_";
        private const string RetailPtrFolderName = "_ptr_";
        private const string BetaFolderName = "_beta_";
        private const string InterfaceFolderName = "Interface";
        private const string AddonFolderName = "AddOns";
        private const string BetaExecutableName = "WowB.exe";
        private const string RetailExecutableName = "Wow.exe";
        private const string RetailPtrExecutableName = "WowT.exe";
        private const string ClassicExecutableName = "WowClassic.exe";
        private const string ClassicPtrExecutableName = "WowClassicT.exe";

        // BLIZZARD STRINGS
        private const string BlizzardAgentFolderFormat = "ProgramData\\Battle.net\\Agent";
        private const string BlizzardProductDbName = "product.db";

        // PREFERENCE KEYS
        private const string WowRetailLocationPreferenceKey = "wow_retail_location";
        private const string WowRetailPtrLocationPreferenceKey = "wow_retail_ptr_location";
        private const string WowClassicLocationPreferenceKey = "wow_classic_location";
        private const string WowClassicPtrLocationPreferenceKey = "wow_classic_ptr_location";
        private const string WowBetaLocationPreferenceKey = "wow_beta_location";

        private readonly IPreferenceRepository _preferenceRepository;

        public event WarcraftEventHandler ProductChanged;

        public IList<InstalledProduct> InstalledProducts { get; private set; }

        public string ProductsDbPath => Path.Combine(GetBlizzardAgentPath(), BlizzardProductDbName);

        public WarcraftService(
            IPreferenceRepository preferenceRepository)
        {
            _preferenceRepository = preferenceRepository;
            SetDefaultPreferences();

            ScanProducts();
        }

        public IList<InstalledProduct> ScanProducts()
        {
            InstalledProducts = DecodeProducts(ProductsDbPath);

            var clientTypes = EnumExtensions.Values<WowClientType>();
            foreach (var clientType in clientTypes)
            {
                var clientLocation = GetClientLocation(clientType);
                var productLocation = GetProductLocation(clientType);

                if(string.IsNullOrEmpty(clientLocation) && string.IsNullOrEmpty(productLocation))
                {
                    continue;
                }

                Log.Information($"clientLocation {clientLocation}, productLocation: {productLocation}");
                if (AreEqualPaths(clientLocation, productLocation))
                {
                    continue;
                }

                // If the path that the user selected is valid, then move on.
                if (!string.IsNullOrEmpty(clientLocation) && IsClientFolder(clientType, clientLocation))
                {
                    continue;
                }

                var locationPreference = GetClientLocationPreference(clientType);
                if (locationPreference == null)
                {
                    locationPreference = _preferenceRepository.Create(GetClientLocationPreferenceKey(clientType), string.Empty);
                }

                locationPreference.Value = productLocation;

                Log.Information($"locationPreference {locationPreference}");

                _preferenceRepository.SaveItem(locationPreference);

                var changedClient = new ChangedClient
                {
                    ClientType = clientType,
                    NewLocation = productLocation,
                    PreviousLocation = clientLocation
                };

                ProductChanged?.Invoke(this, new WarcraftEventArgs
                {
                    ChangedClient = changedClient
                });
            }

            return InstalledProducts;
        }

        private bool AreEqualPaths(string path1, string path2)
        {
            if (string.IsNullOrEmpty(path1) && string.IsNullOrEmpty(path2))
            {
                return true;
            }
            if (string.IsNullOrEmpty(path1) && !string.IsNullOrEmpty(path2))
            {
                return false;
            }
            if (string.IsNullOrEmpty(path2) && !string.IsNullOrEmpty(path1))
            {
                return false;
            }

            return Path.GetFullPath(path1) == Path.GetFullPath(path2);
        }

        public IList<string> GetWowClientNames()
        {
            var clients = GetWowClientTypes();
            return clients
                .Select(c => c.GetDisplayName())
                .ToList();
        }

        public string GetClientLocation(WowClientType clientType)
        {
            var preference = GetClientLocationPreference(clientType);
            return preference?.Value ?? string.Empty;
        }

        private Preference GetClientLocationPreference(WowClientType clientType)
        {
            var preferenceKey = GetClientLocationPreferenceKey(clientType);
            return _preferenceRepository.FindByKey(preferenceKey);
        }

        public IList<string> GetClientLocations()
        {
            var clientTypes = EnumExtensions.Values<WowClientType>();
            return clientTypes.Select(clientType => GetClientLocation(clientType)).ToList();
        }

        public IList<WowClientType> GetWowClientTypes()
        {
            IList<WowClientType> clients = new List<WowClientType>();

            var clientTypes = EnumExtensions.Values<WowClientType>();
            foreach (var clientType in clientTypes)
            {
                var clientLocation = GetClientLocation(clientType);
                if (string.IsNullOrEmpty(clientLocation) || !Directory.Exists(clientLocation))
                {
                    continue;
                }

                clients.Add(clientType);
            }

            return clients;
        }

        public bool IsClientFolder(WowClientType clientType, string folderPath)
        {
            var clientFolderName = GetClientFolderName(clientType);
            var executableName = GetExecuteableName(clientType);
            var executablePath = Path.Combine(folderPath, clientFolderName, executableName);

            return File.Exists(executablePath);
        }

        public string GetAddonFolderPath(WowClientType clientType)
        {
            var wowPath = GetClientLocation(clientType);
            var clientFolder = GetClientFolderName(clientType);
            var executableFolder = Path.Combine(wowPath, clientFolder);

            if (!Directory.Exists(executableFolder))
            {
                throw new Exception($"Wow client folder not found {executableFolder}");
            }

            return Path.Combine(executableFolder, InterfaceFolderName, AddonFolderName);
        }

        public bool SetWowFolderPath(WowClientType clientType, string folderPath)
        {
            if (!IsClientFolder(clientType, folderPath))
            {
                return false;
            }

            var preferenceKey = GetClientLocationPreferenceKey(clientType);
            var preference = _preferenceRepository.FindByKey(preferenceKey);
            if (preference == null)
            {
                throw new Exception("client preference key not found");
            }

            preference.Value = folderPath;
            _preferenceRepository.UpdateItem(preference);

            return true;
        }

        public async Task<IEnumerable<AddonFolder>> ListAddons(WowClientType clientType)
        {
            if(clientType == WowClientType.None)
            {
                return new List<AddonFolder>();
            }

            var addons = new List<AddonFolder>();

            var addonsPath = GetAddonFolderPath(clientType);
            var addonsDirectory = new DirectoryInfo(addonsPath);

            // Folder may not exist if no addons have been installed
            if (!addonsDirectory.Exists)
            {
                return addons;
            }

            var addonDirectories = addonsDirectory.GetDirectories();
            Log.Debug($"addonDirectories {addonDirectories.Length}");

            foreach (var directory in addonDirectories)
            {
                Log.Debug($"Getting addon folder {directory.Name}");
                var addonFolder = await GetAddonFolder(directory);
                addons.Add(addonFolder);
            }

            return addons;
        }

        public string GetClientFolderName(WowClientType clientType)
        {
            return clientType switch
            {
                WowClientType.Retail => RetailFolderName,
                WowClientType.Classic => ClassicFolderName,
                WowClientType.RetailPtr => RetailPtrFolderName,
                WowClientType.ClassicPtr => ClassicPtrFolderName,
                WowClientType.Beta => BetaFolderName,
                _ => string.Empty,
            };
        }

        private async Task<AddonFolder> GetAddonFolder(DirectoryInfo directory)
        {
            var toc = await ParseToc(directory);
            //var tocMetaData = await ParseTocMetadata(directory);

            return new AddonFolder
            {
                Directory = directory,
                Name = directory.Name,
                Path = directory.FullName,
                Status = "Pending",
                Toc = toc,
                //TocMetaData = tocMetaData
            };
        }

        private async Task<Toc> ParseToc(DirectoryInfo directory)
        {
            var files = directory.GetFiles();
            var tocFile = files.FirstOrDefault(f => f.Extension.Contains("toc"));
            if (tocFile == null)
            {
                return default;
            }

            var fileText = await FileUtilities.GetFileTextAsync(tocFile.FullName);
            return new TocParser(fileText).Toc;
        }

        private async Task<IList<string>> ParseTocMetadata(DirectoryInfo directory)
        {
            var files = directory.GetFiles();
            var tocFile = files.FirstOrDefault(f => f.Extension.Contains("toc"));
            if (tocFile == null)
            {
                return default;
            }

            var fileText = await FileUtilities.GetFileTextAsync(tocFile.FullName);
            return new TocParser(fileText).GetMetaData();
        }

        private string GetClientLocationPreferenceKey(WowClientType clientType)
        {
            return clientType switch
            {
                WowClientType.Retail => WowRetailLocationPreferenceKey,
                WowClientType.Classic => WowClassicLocationPreferenceKey,
                WowClientType.RetailPtr => WowRetailPtrLocationPreferenceKey,
                WowClientType.ClassicPtr => WowClassicPtrLocationPreferenceKey,
                WowClientType.Beta => WowBetaLocationPreferenceKey,
                _ => string.Empty,
            };
        }

        private string GetExecuteableName(WowClientType clientType)
        {
            return clientType switch
            {
                WowClientType.Retail => RetailExecutableName,
                WowClientType.Classic => ClassicExecutableName,
                WowClientType.RetailPtr => RetailPtrExecutableName,
                WowClientType.ClassicPtr => ClassicPtrExecutableName,
                WowClientType.Beta => BetaExecutableName,
                _ => string.Empty,
            };
        }

        private string GetBlizzardAgentPath()
        {
            var driveLetters = FileUtilities.GetAllDriveLetters();
            Log.Debug($"Drives: {string.Join(",", driveLetters)}");

            foreach (var drive in driveLetters)
            {
                var agentPath = Path.Combine(drive, BlizzardAgentFolderFormat);
                Log.Debug($"Searching for blizzard agent: {agentPath}");

                if (Directory.Exists(agentPath))
                {
                    return agentPath;
                }
            }

            Log.Debug("No blizzard agent path was found");
            return string.Empty;
        }

        private IList<InstalledProduct> DecodeProducts(string productDbPath)
        {
            var products = new List<InstalledProduct>();

            try
            {
                using var fileStream = File.OpenRead(productDbPath);
                var productDb = Serializer.Deserialize<ProductDb>(fileStream);

                var installedProducts = productDb.Products
                    .Where(p => p.Family == "wow")
                    .Select(p => new InstalledProduct
                    {
                        Location = p.Client.Location,
                        Name = p.Client.Name
                    });

                products.AddRange(installedProducts);
            }
            catch (Exception ex)
            {
                Log.Error(ex, "Failed to decode product db");
            }

            return products;
        }

        private string GetProductLocation(WowClientType clientType)
        {
            var clientFolderName = GetClientFolderName(clientType);

            var clientLocation = InstalledProducts.FirstOrDefault(p => p.Name == clientFolderName);

            return clientLocation?.Location ?? string.Empty;
        }

        private void SetDefaultPreferences()
        {
            var retailLocation = _preferenceRepository.FindByKey(WowRetailLocationPreferenceKey);
            var retailPtrLocation = _preferenceRepository.FindByKey(WowRetailPtrLocationPreferenceKey);
            var classicLocation = _preferenceRepository.FindByKey(WowClassicLocationPreferenceKey);
            var classicPtrLocation = _preferenceRepository.FindByKey(WowClassicPtrLocationPreferenceKey);

            if (retailLocation == null)
            {
                _preferenceRepository.Create(WowRetailLocationPreferenceKey, string.Empty);
            }

            if (retailPtrLocation == null)
            {
                _preferenceRepository.Create(WowRetailPtrLocationPreferenceKey, string.Empty);
            }

            if (classicLocation == null)
            {
                _preferenceRepository.Create(WowClassicLocationPreferenceKey, string.Empty);
            }

            if (classicPtrLocation == null)
            {
                _preferenceRepository.Create(WowClassicPtrLocationPreferenceKey, string.Empty);
            }
        }
    }
}

File: wowup-electron/src/app/core/services/warcraft/warcraft.service.ts
import { Injectable } from "@angular/core";
import { WarcraftServiceImpl } from "./warcraft.service.impl";
import { WarcraftServiceWin } from "./warcraft.service.win";
import { FileUtils } from "app/core/utils/file.utils";
import { ProductDb } from "app/models/warcraft/product-db";
import { InstalledProduct } from "app/models/warcraft/installed-product";
import { from, BehaviorSubject } from "rxjs";
import * as path from 'path';
import * as fs from 'fs';
import { promisify } from 'util';
import { map, filter, delay } from "rxjs/operators";
import { WowClientType } from "app/models/warcraft/wow-client-type";
import { StorageService } from "../storage/storage.service";
import log from 'electron-log';
import { WarcraftServiceMac } from "./warcraft.service.mac";
import { AddonFolder } from "app/models/wowup/addon-folder";
import { TocParser } from "app/core/utils/toc-parser";

const isWin = process.platform === "win32";
const isMac = process.platform === "darwin";

// WOW STRINGS
const CLIENT_RETAIL_FOLDER = '_retail_';
const CLIENT_RETAIL_PTR_FOLDER = '_ptr_';
const CLIENT_CLASSIC_FOLDER = '_classic_';
const CLIENT_CLASSIC_PTR_FOLDER = '_classic_ptr_';
const CLIENT_BETA_FOLDER = '_beta_';
const ADDON_FOLDER_NAME = 'AddOns';
const INTERFACE_FOLDER_NAME = 'Interface';

// PREFERENCE KEYS
const RETAIL_LOCATION_KEY = "wow_retail_location";
const RETAIL_PTR_LOCATION_KEY = "wow_retail_ptr_location";
const CLASSIC_LOCATION_KEY = "wow_classic_location";
const CLASSIC_PTR_LOCATION_KEY = "wow_classic_ptr_location";
const BETA_LOCATION_KEY = "wow_beta_location";

const readdirAsync = promisify(fs.readdir);

@Injectable({
    providedIn: 'root'
})
export class WarcraftService {

    private readonly _impl: WarcraftServiceImpl;
    private readonly _productsSrc = new BehaviorSubject<InstalledProduct[]>([]);

    private _productDbPath = '';

    public products$ = this._productsSrc.asObservable();
    public productsReady$ = this.products$
        .pipe(
            filter(products => Array.isArray(products))
        );

    public clientTypes$ = this.products$
        .pipe(
            map(products => products.map(product => product.clientType))
        );

    public clientTypesNames$ = this.products$
        .pipe(
            map(products => products.map(product => this.getClientDisplayName(product.clientType)))
        );

    constructor(
        private storage: StorageService
    ) {
        this._impl = this.getImplementation();

        from(this._impl.getBlizzardAgentPath())
            .pipe(
                map(productDbPath => this._productDbPath = productDbPath),
                map(() => this.scanProducts())
            )
            .subscribe();
    }

    public getExecutableName(clientType: WowClientType): string {
        return this._impl.getExecutableName(clientType);
    }

    public getFullExecutablePath(clientType: WowClientType): string {
        const clientLocation = this.getClientLocation(clientType);
        const clientFolder = this.getClientFolderName(clientType);
        const clientExecutable = this.getExecutableName(clientType);

        return path.join(clientLocation, clientFolder, clientExecutable);
    }

    public getFullClientPath(clientType: WowClientType): string {
        const clientLocation = this.getClientLocation(clientType);
        const clientFolder = this.getClientFolderName(clientType);

        return path.join(clientLocation, clientFolder);
    }

    public getAddonFolderPath(clientType: WowClientType): string {
        const fullClientPath = this.getFullClientPath(clientType);
        return path.join(fullClientPath, INTERFACE_FOLDER_NAME, ADDON_FOLDER_NAME);
    }

    public scanProducts() {
        const installedProducts = this.decodeProducts(this._productDbPath);

        for (let product of installedProducts) {
            const clientType = product.clientType;
            const clientLocation = this.getClientLocation(clientType);

            if (this.arePathsEqual(clientLocation, product.location)) {
                continue;
            }

            this.setClientLocation(clientType, product.location);
        }

        this._productsSrc.next(installedProducts);

        return installedProducts;
    }

    public async listAddons(clientType: WowClientType) {
        const addonFolders: AddonFolder[] = [];
        const addonFolderPath = this.getAddonFolderPath(clientType);

        // Folder may not exist if no addons have been installed
        if (!fs.existsSync(addonFolderPath)) {
            return addonFolders;
        }

        const files = fs.readdirSync(addonFolderPath, { withFileTypes: true });
        const directories = files.filter(dirent => dirent.isDirectory()).map(dirent => dirent.name);

        for (let i = 0; i < directories.length; i += 1) {
            const dir = directories[i];
            const addonFolder = await this.getAddonFolder(addonFolderPath, dir);
            if (addonFolder) {
                addonFolders.push(addonFolder);
            }
        }

        return addonFolders;
    }

    private async getAddonFolder(addonFolderPath: string, dir: string): Promise<AddonFolder> {
        try {
            const dirPath = path.join(addonFolderPath, dir);
            const dirFiles = fs.readdirSync(dirPath);
            // const dirFiles = await readdirAsync(dirPath);
            const tocFile = dirFiles.find(f => path.extname(f) === '.toc');
            if (!tocFile) {
                return null;
            }

            const toc = await this.getToc(path.join(dirPath, tocFile));

            return {
                name: dir,
                path: dirPath,
                status: 'Pending',
                toc: toc
            };
        } catch (e) {
            console.error(e);
            return null;
        }
    }

    private async getToc(filePath: string) {
        const parser = new TocParser(filePath);
        return await parser.parse();
    }

    public getClientLocation(clientType: WowClientType) {
        const clientLocationKey = this.getClientLocationKey(clientType);
        return this.storage.getPreference<string>(clientLocationKey) || '';
    }

    public setClientLocation(clientType: WowClientType, clientPath: string) {
        const clientLocationKey = this.getClientLocationKey(clientType);
        return this.storage.setPreference(clientLocationKey, clientPath);
    }

    public getClientTypeForFolderName(folderName: string): WowClientType {
        switch (folderName) {
            case CLIENT_RETAIL_FOLDER:
                return WowClientType.Retail;
            case CLIENT_RETAIL_PTR_FOLDER:
                return WowClientType.RetailPtr;
            case CLIENT_CLASSIC_FOLDER:
                return WowClientType.Classic;
            case CLIENT_CLASSIC_PTR_FOLDER:
                return WowClientType.ClassicPtr;
            case CLIENT_BETA_FOLDER:
                return WowClientType.Beta;
            default:
                return WowClientType.Retail;
        }
    }

    public getClientFolderName(clientType: WowClientType) {
        switch (clientType) {
            case WowClientType.Retail:
                return CLIENT_RETAIL_FOLDER;
            case WowClientType.Classic:
                return CLIENT_CLASSIC_FOLDER;
            case WowClientType.RetailPtr:
                return CLIENT_RETAIL_PTR_FOLDER;
            case WowClientType.ClassicPtr:
                return CLIENT_CLASSIC_PTR_FOLDER;
            case WowClientType.Beta:
                return CLIENT_BETA_FOLDER;
            default:
                return '';
        }
    }

    public getClientDisplayName(clientType: WowClientType) {
        switch (clientType) {
            case WowClientType.Retail:
                return 'Retail';
            case WowClientType.Classic:
                return 'Classic';
            case WowClientType.RetailPtr:
                return 'Retail PTR';
            case WowClientType.ClassicPtr:
                return 'Classic PTR';
            case WowClientType.Beta:
                return 'Beta';
            default:
                return '';
        }
    }

    public getClientLocationKey(clientType: WowClientType) {
        switch (clientType) {
            case WowClientType.Retail:
                return RETAIL_LOCATION_KEY;
            case WowClientType.Classic:
                return CLASSIC_LOCATION_KEY;
            case WowClientType.RetailPtr:
                return RETAIL_PTR_LOCATION_KEY;
            case WowClientType.ClassicPtr:
                return CLASSIC_PTR_LOCATION_KEY;
            case WowClientType.Beta:
                return BETA_LOCATION_KEY;
            default:
                return '';
        }
    }

    private decodeProducts(productDbPath: string) {
        const productDbData = FileUtils.readFileSync(productDbPath);

        try {
            const productDb = ProductDb.decode(productDbData);
            const wowProducts: InstalledProduct[] = productDb.products
                .filter(p => p.family === 'wow')
                .map(p => ({
                    location: p.client.location,
                    name: p.client.name,
                    clientType: this.getClientTypeForFolderName(p.client.name)
                }));

            console.log(wowProducts)
            return wowProducts;
        } catch (e) {
            console.error('failed to decode product db')
            console.error(e);
            return [];
        }
    }

    private arePathsEqual(path1: string, path2: string) {
        return path.normalize(path1) === path.normalize(path2);
    }

    private getImplementation(): WarcraftServiceImpl {
        if (isWin) {
            return new WarcraftServiceWin();
        }

        if (isMac) {
            return new WarcraftServiceMac();
        }

        throw new Error('No warcraft service implementation found');
    }
}